File: backend/app/models/action_log.py
from datetime import datetime
from sqlalchemy import String, ForeignKey, DateTime, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from .base import Base


class ActionLog(Base):
    __tablename__ = "action_logs"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    device_id: Mapped[int] = mapped_column(ForeignKey("devices.id"), nullable=False)
    action: Mapped[str] = mapped_column(String(100), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user: Mapped["User | None"] = relationship(back_populates="action_logs")
    device: Mapped["Device"] = relationship(back_populates="action_logs")

File: mobile/lib/models/action_log.dart
class ActionLog {
  final int id;
  final int? userId;
  final int deviceId;
  final String action;
  final DateTime createdAt;

  const ActionLog({
    required this.id,
    required this.userId,
    required this.deviceId,
    required this.action,
    required this.createdAt,
  });

  factory ActionLog.fromJson(Map<String, dynamic> json) => ActionLog(
        id: json['id'] as int,
        userId: json['user_id'] as int?,
        deviceId: json['device_id'] as int,
        action: json['action'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}
